src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala

```diff
@@ -47,25 +47,18 @@ class InstanceOpFactoryImpl(
   override def matchOfferRequest(request: InstanceOpFactory.Request): OfferMatchResult = {
     logger.debug("matchOfferRequest")
 
-    request.runSpec match {
-      case app: AppDefinition =>
-        if (request.isForResidentRunSpec) {
-          inferForResidents(request)
-        } else {
-          request.scheduledInstances.headOption.map { scheduledInstance =>
-            inferNormalTaskOp(app, request.instances, request.offer, request.localRegion, scheduledInstance)
-          }.getOrElse(OfferMatchResult.NoMatch(app, request.offer, Seq.empty, clock.now()))
-        }
-      case pod: PodDefinition =>
-        if (request.isForResidentRunSpec) {
-          inferForResidents(request)
-        } else {
-          request.scheduledInstances.headOption.map { scheduledInstance =>
-            inferPodInstanceOp(pod, request.instances, request.offer, request.localRegion, scheduledInstance)
-          }.getOrElse(OfferMatchResult.NoMatch(pod, request.offer, Seq.empty, clock.now()))
-        }
-      case _ =>
-        throw new IllegalArgumentException(s"unsupported runSpec object ${request.runSpec}")
+    request.scheduledInstances.headOption match {
+      case Some(scheduledInstance @ Instance(_, _, _, _, app: AppDefinition, _)) =>
+        if (app.isResident) inferForResidents(request, scheduledInstance)
+        else inferNormalTaskOp(app, request.instances, request.offer, request.localRegion, scheduledInstance)
+      case Some(scheduledInstance @ Instance(_, _, _, _, pod: PodDefinition, _)) =>
+        if (pod.isResident) inferForResidents(request, scheduledInstance)
+        inferPodInstanceOp(pod, request.instances, request.offer, request.localRegion, scheduledInstance)
+      case Some(Instance(_, _, _, _, runSpec, _)) =>
+        throw new IllegalArgumentException(s"unsupported runSpec object ${runSpec}")
+      case None =>
+        // TODO(karsten): Ideally scheduled instances would be a non empty seq
+        OfferMatchResult.NoScheduledInstances(request.offer, clock.now())
```

Comment:
Cats offers a `NonEmptyList` are we brace enough to import it or should we have an `IllegalStateException` instead? There should be no offer matching when no instance is scheduled.

Reply:
I agree - NonEmptyList would make sense.